java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java

```diff
@@ -319,4 +321,73 @@ static void setBitMaskedByte(ArrowBuf data, int byteIndex, byte bitMask) {
     currentByte |= bitMask;
     data.setByte(byteIndex, currentByte);
   }
+
+  /**
+   * Concat two validity buffers.
+   * @param input1 the first validity buffer.
+   * @param numBits1 the number of bits in the first validity buffer.
+   * @param input2 the second validity buffer.
+   * @param numBits2 the number of bits in the second validity buffer.
+   * @param output the ouput validity buffer. It can be the same one as the first input.
+   */
+  public static void concatBits(ArrowBuf input1, int numBits1, ArrowBuf input2, int numBits2, ArrowBuf output) {
+    int numBytes1 = DataSizeRoundingUtil.divideBy8Ceil(numBits1);
+    int numBytes2 = DataSizeRoundingUtil.divideBy8Ceil(numBits2);
+    int numBytesOut = DataSizeRoundingUtil.divideBy8Ceil(numBits1 + numBits2);
+
+    if (BoundsChecking.BOUNDS_CHECKING_ENABLED) {
+      output.checkBytes(0, numBytesOut);
+    }
+
+    // copy the first bit set
+    if (input1 != output) {
+      PlatformDependent.copyMemory(input1.memoryAddress(), output.memoryAddress(), numBytes1);
+    }
+    output.setZero(numBytes1, numBytesOut);
+
+    if (numBits1 % 8 == 0) {
+      PlatformDependent.copyMemory(input2.memoryAddress(), output.memoryAddress() + numBytes1, numBytes2);
+      return;
+    }
+
+    // the number of bits to fill a full byte after the first input is processed
+    int numBitsToFill = 8 - (numBits1 % 8);
+
+    // the number of extra bits for the second input, relative to full bytes
+    int numRemainingBits = numBits2 % 8;
+
+    int numFullBytes = numBits2 / 8;
+
+    for (int i = 0; i < numFullBytes; i++) {
+      int prevByte = output.getByte(numBytes1 + i - 1) & 0xff;
+      byte curByte = input2.getByte(i);
+
+      // first fill the bits to a full byte
+      int byteToFill = (curByte & 0xff) << (8 - numBitsToFill);
+      output.setByte(numBytes1 + i - 1, byteToFill | prevByte);
+
+      // fill remaining bits in the current byte
+      int remByte = (curByte & 0xff) >>> numBitsToFill;
+      output.setByte(numBytes1 + i, remByte);
```

Comment:
it seems like you could avoid one memory access, by keeping this byte cached instead of accessing writing it here and then reading it back again at the beginning of the loop.

Reply:
Good point. Revised accordingly. Thank you.

Reply:
it still seems like this is setting memory twice per cycle?

Reply:
Sorry I misunderstood your point at first. I have reivsed the code accordingly. Thanks for the great suggestion.